app/maptools/recordingmaptool.cpp

```diff
@@ -406,3 +401,94 @@ void RecordingMapTool::setHandles( const QgsGeometry &newHandles )
   mHandles = newHandles;
   emit handlesChanged( mHandles );
 }
+
+const QString &RecordingMapTool::state() const
+{
+  return mState;
+}
+
+void RecordingMapTool::setState( const QString &newState )
+{
+  if ( mState == newState )
+    return;
+  mState = newState;
+  emit stateChanged( mState );
+}
+
+QgsVertexId &RecordingMapTool::clickedVertexId()
+{
+  return mClickedVertexId;
+}
+
+void RecordingMapTool::setClickedVertexId( QgsVertexId newId )
+{
+  if ( mClickedVertexId == newId )
+    return;
+  mClickedVertexId.part = newId.part;
+  mClickedVertexId.ring = newId.ring;
+  mClickedVertexId.vertex = newId.vertex;
+  emit clickedVertexIdChanged( mClickedVertexId );
+}
+
+void RecordingMapTool::lookForVertex( const QPointF &clickedPoint, double searchRadius )
+{
+  double minDistance = std::numeric_limits<double>::max();
+  double currentDistance = 0;
+  QgsPoint point;
+  QgsVertexId vertexId;
+
+  QgsPoint pnt = mapSettings()->screenToCoordinate( clickedPoint );
+
+  if ( mInitialGeometry.isEmpty() )
+  {
+    setClickedVertexId( vertexId );
+    return;
+  }
+
+  for ( QPair<QgsVertexId, QgsPoint> pair : mVertexIds )
+  {
+    currentDistance = QgsGeometryUtils::sqrDistance2D( pnt, pair.second );
+    if ( currentDistance <= minDistance && currentDistance <= searchRadius )
```

Comment:
Why do we need the `minDistance`?

Reply:
We discussed it a few days ago. If user taps somewhere too far from the any vertex we still will be able to find closest vertex based on the distance to the clicked position. So we want to search only within some radius around each vertex.